optaplanner-core/src/main/java/org/optaplanner/core/impl/score/inliner/JustificationsSupplier.java

```diff
@@ -14,20 +14,19 @@
  * limitations under the License.
  */
 
-package org.optaplanner.core.impl.score.stream;
+package org.optaplanner.core.impl.score.inliner;
 
-import org.optaplanner.core.api.score.Score;
-import org.optaplanner.core.impl.score.director.InnerScoreDirector;
-
-public interface ConstraintSessionFactory<Solution_, Score_ extends Score<Score_>> {
-
-    /**
-     * This method is thread-safe.
-     *
-     * @param constraintMatchEnabled true if {@link InnerScoreDirector#isConstraintMatchEnabled()} should be true
-     * @param workingSolution if null, uniform synthetic constraint weights will be applied
-     * @return never null
-     */
-    Object buildSession(boolean constraintMatchEnabled, Solution_ workingSolution);
+import java.util.List;
+import java.util.function.Supplier;
 
+/**
+ * This interface exists so that justifications can be created lazily
+ * and for that pattern to naturally spread throughout the codebase.
+ *
+ * Justifications must be specifically enabled, and that is usually done outside of the hot path,
+ * as that code is expensive.
+ * Therefore constructing lists of justifications in case they are ever needed is a waste of CPU cycles.
```

Comment:
To bring it in sync with the changes above:

```suggestion
 * Justifications creation is performance expensive and constraint matches are typically disabled.
 * So justifications are created lazily, outside of the typical hot path.
```

(all soft suggestions of course)